File: src/main/java/meowing/zen/featManager.java
package meowing.zen;

import meowing.zen.config.zencfg;
import meowing.zen.utils.EventBus;
import java.util.*;
import java.util.concurrent.ConcurrentHashMap;
import java.util.concurrent.atomic.AtomicInteger;
import java.util.logging.Logger;
import java.nio.file.*;
import java.lang.reflect.Field;
import java.io.IOException;

public class featManager {
    private static final Logger LOGGER = Logger.getLogger("Zen");
    private static final Map<String, FeatureHandler> features = new ConcurrentHashMap<>();
    private static final Map<String, Boolean> configState = new ConcurrentHashMap<>();
    private static final Map<String, Field> fieldCache = new ConcurrentHashMap<>();
    private static final String featpath = "/meowing/zen/feats";
    private static final Set<String> immutableFeats = ConcurrentHashMap.newKeySet();
    private static final AtomicInteger moduleCount = new AtomicInteger(0);
    
    static {
        cacheConf();
    }
    
    private static void cacheConf() {
        Field[] fields = zencfg.class.getDeclaredFields();
        for (Field field : fields) {
            if (field.getType() == boolean.class) {
                field.setAccessible(true);
                fieldCache.put(field.getName(), field);
            }
        }
    }
    
    private static class FeatureHandler {
        final Object owner;
        final Runnable registerHandler;
        volatile boolean isActive = false;
        
        FeatureHandler(Object owner, Runnable registerHandler) {
            this.owner = owner;
            this.registerHandler = registerHandler;
        }
        
        boolean toggle(boolean active) {
            if (active && !isActive) {
                try {
                    registerHandler.run();
                    isActive = true;
                    return true;
                } catch (Exception e) {
                    LOGGER.warning("Failed to activate feature: " + owner.getClass().getSimpleName() + " - " + e.getMessage());
                    return false;
                }
            } else if (!active && isActive) {
                try {
                    EventBus.unregister(owner);
                    isActive = false;
                    return true;
                } catch (Exception e) {
                    LOGGER.warning("Failed to deactivate feature: " + owner.getClass().getSimpleName() + " - " + e.getMessage());
                    return false;
                }
            }
            return false;
        }
    }
    
    private static void autoDiscFeats() {
        var resource = featManager.class.getResource(featpath);
        if (resource == null) {
            LOGGER.warning("Features directory not found: " + featpath);
            return;
        }
        
        try {
            var uri = resource.toURI();
            if (uri.getScheme().equals("jar")) {
                try (var fs = getOrCreateFileSystem(uri)) {
                    discoverFeats(fs.getPath(featpath));
                }
            } else {
                discoverFeats(Paths.get(uri));
            }
        } catch (Exception e) {
            LOGGER.severe("Failed to discover features: " + e.getMessage());
        }
    }
    
    private static FileSystem getOrCreateFileSystem(java.net.URI uri) throws IOException {
        try {
            return FileSystems.getFileSystem(uri);
        } catch (FileSystemNotFoundException e) {
            return FileSystems.newFileSystem(uri, Collections.emptyMap());
        }
    }
    
    private static void discoverFeats(Path path) {
        try (var stream = Files.walk(path, 3)) {
            stream.filter(p -> {
                String pathStr = p.toString();
                return pathStr.endsWith(".class") && pathStr.contains("meowing/zen/feats");
            })
            .parallel()
            .forEach(featManager::loadFeats);
        } catch (IOException e) {
            LOGGER.severe("Failed to walk features directory: " + e.getMessage());
        }
    }
    
    private static void loadFeats(Path classPath) {
        try {
            String pathStr = classPath.toString();
            int startIdx = pathStr.indexOf("meowing/zen/feats");
            if (startIdx == -1) return;
            
            String className = pathStr.substring(startIdx)
                .replace('/', '.')
                .replace('\\', '.')
                .replace(".class", "");
            
            Class<?> clazz = Class.forName(className);
            var initMethod = clazz.getDeclaredMethod("initialize");
            initMethod.invoke(null);
            
            moduleCount.incrementAndGet();
        } catch (ReflectiveOperationException e) {
            LOGGER.fine("Skipping non-feature class or initialization failed: " + classPath);
        } catch (Exception e) {
            LOGGER.warning("Unexpected error loading feature class " + classPath + ": " + e.getMessage());
        }
    }
    
    public static void register(Object owner, Runnable registerHandler) {
        Objects.requireNonNull(owner, "Feature owner cannot be null");
        Objects.requireNonNull(registerHandler, "Register handler cannot be null");
        
        String name = owner.getClass().getSimpleName().toLowerCase();
        var handler = new FeatureHandler(owner, registerHandler);
        
        features.put(name, handler);
        immutableFeats.add(name);
        boolean shouldBeActive = getConfigValue(name);
        configState.put(name, shouldBeActive);
        handler.toggle(shouldBeActive);
    }
    
    public static void onConfigChange() {
        features.entrySet().parallelStream().forEach(entry -> {
            String name = entry.getKey();
            var handler = entry.getValue();
            boolean shouldBeActive = getConfigValue(name);
            Boolean lastValue = configState.get(name);
            
            if (lastValue == null || lastValue != shouldBeActive) 
                if (handler.toggle(shouldBeActive)) configState.put(name, shouldBeActive);
        });
    }
    
    private static boolean getConfigValue(String fieldName) {
        Field field = fieldCache.get(fieldName);
        if (field == null) return false;
        
        try {
            return (Boolean) field.get(Zen.getConfig());
        } catch (IllegalAccessException e) {
            LOGGER.warning("Failed to access config field: " + fieldName);
            return false;
        } catch (ClassCastException e) {
            LOGGER.warning("Config field is not boolean: " + fieldName);
            return false;
        }
    }
    
    public static void initAll() {
        autoDiscFeats();
    }
    
    public static int getFeatCount() {
        return (int) features.values().stream()
            .mapToInt(handler -> handler.isActive ? 1 : 0)
            .sum();
    }
    
    public static Set<String> getRegFeats() {
        return immutableFeats;
    }
    
    public static int getModuleCount() {
        return moduleCount.get();
    }
}

File: src/main/kotlin/meowing/zen/featManager.kt
package meowing.zen

import meowing.zen.config.zencfg
import meowing.zen.utils.EventBus
import java.util.*
import java.util.concurrent.ConcurrentHashMap
import java.util.concurrent.atomic.AtomicInteger
import java.util.logging.Logger
import java.nio.file.*
import java.lang.reflect.Field
import java.io.IOException

object featManager {
    private val LOGGER = Logger.getLogger("Zen")
    private val features = ConcurrentHashMap<String, FeatureHandler>()
    private val configState = ConcurrentHashMap<String, Boolean>()
    private val fieldCache = ConcurrentHashMap<String, Field>()
    private const val featpath = "/meowing/zen/feats"
    private val immutableFeats = ConcurrentHashMap.newKeySet<String>()
    private val moduleCount = AtomicInteger(0)

    init {
        cacheConf()
    }

    private fun cacheConf() {
        val fields = zencfg::class.java.declaredFields
        for (field in fields) {
            if (field.type == Boolean::class.javaPrimitiveType) {
                field.isAccessible = true
                fieldCache[field.name] = field
            }
        }
    }

    private class FeatureHandler(
        val owner: Any,
        val registerHandler: Runnable
    ) {
        @Volatile
        var isActive = false

        fun toggle(active: Boolean): Boolean {
            return when {
                active && !isActive -> {
                    try {
                        registerHandler.run()
                        isActive = true
                        true
                    } catch (e: Exception) {
                        LOGGER.warning("Failed to activate feature: ${owner::class.java.simpleName} - ${e.message}")
                        false
                    }
                }
                !active && isActive -> {
                    try {
                        EventBus.unregister(owner)
                        isActive = false
                        true
                    } catch (e: Exception) {
                        LOGGER.warning("Failed to deactivate feature: ${owner::class.java.simpleName} - ${e.message}")
                        false
                    }
                }
                else -> false
            }
        }
    }

    private fun autoDiscFeats() {
        val resource = featManager::class.java.getResource(featpath)
        if (resource == null) {
            LOGGER.warning("Features directory not found: $featpath")
            return
        }

        try {
            val uri = resource.toURI()
            if (uri.scheme == "jar") {
                getOrCreateFileSystem(uri).use { fs ->
                    discoverFeats(fs.getPath(featpath))
                }
            } else {
                discoverFeats(Paths.get(uri))
            }
        } catch (e: Exception) {
            LOGGER.severe("Failed to discover features: ${e.message}")
        }
    }

    private fun getOrCreateFileSystem(uri: java.net.URI): FileSystem {
        return try {
            FileSystems.getFileSystem(uri)
        } catch (e: FileSystemNotFoundException) {
            FileSystems.newFileSystem(uri, emptyMap<String, Any>())
        }
    }

    private fun discoverFeats(path: Path) {
        try {
            Files.walk(path, 3).use { stream ->
                stream.filter { p ->
                    val pathStr = p.toString()
                    pathStr.endsWith(".class") && pathStr.contains("meowing/zen/feats")
                }
                    .parallel()
                    .forEach(::loadFeats)
            }
        } catch (e: IOException) {
            LOGGER.severe("Failed to walk features directory: ${e.message}")
        }
    }

    private fun loadFeats(classPath: Path) {
        try {
            val pathStr = classPath.toString()
            val startIdx = pathStr.indexOf("meowing/zen/feats")
            if (startIdx == -1) return

            val className = pathStr.substring(startIdx)
                .replace('/', '.')
                .replace('\\', '.')
                .replace(".class", "")

            val clazz = Class.forName(className)
            val initMethod = clazz.getDeclaredMethod("initialize")
            initMethod.invoke(null)

            moduleCount.incrementAndGet()
        } catch (e: ReflectiveOperationException) {
            LOGGER.fine("Skipping non-feature class or initialization failed: $classPath")
        } catch (e: Exception) {
            LOGGER.warning("Unexpected error loading feature class $classPath: ${e.message}")
        }
    }

    fun register(owner: Any, registerHandler: Runnable) {
        Objects.requireNonNull(owner, "Feature owner cannot be null")
        Objects.requireNonNull(registerHandler, "Register handler cannot be null")

        val name = owner::class.java.simpleName.lowercase()
        val handler = FeatureHandler(owner, registerHandler)

        features[name] = handler
        immutableFeats.add(name)
        val shouldBeActive = getConfigValue(name)
        configState[name] = shouldBeActive
        handler.toggle(shouldBeActive)
    }

    fun onConfigChange() {
        features.entries.parallelStream().forEach { (name, handler) ->
            val shouldBeActive = getConfigValue(name)
            val lastValue = configState[name]

            if (lastValue == null || lastValue != shouldBeActive) {
                if (handler.toggle(shouldBeActive)) {
                    configState[name] = shouldBeActive
                }
            }
        }
    }

    private fun getConfigValue(fieldName: String): Boolean {
        val field = fieldCache[fieldName] ?: return false

        return try {
            field.get(Zen.getConfig()) as Boolean
        } catch (e: IllegalAccessException) {
            LOGGER.warning("Failed to access config field: $fieldName")
            false
        } catch (e: ClassCastException) {
            LOGGER.warning("Config field is not boolean: $fieldName")
            false
        }
    }

    fun initAll() {
        autoDiscFeats()
    }

    fun getFeatCount(): Int {
        return features.values.stream()
            .mapToInt { handler -> if (handler.isActive) 1 else 0 }
            .sum()
    }

    fun getRegFeats(): Set<String> {
        return immutableFeats
    }

    fun getModuleCount(): Int {
        return moduleCount.get()
    }
}